### DISCUSSION TITLE
ulfius_add_endpoint_by_val for files in folder

### DISCUSSION THREAD

**Discussion by nicolasb565** [babelouest/ulfius#232] [General]
Hi, 

I am wondering if there is a way to use ulfius_add_endpoint_by_val for all files inside a folder. For example this would be useful for assets like jpeg files. Can we use wildcard or regex?

Thanks!

**Comment by babelouest** [comment]
Hi @nicolasb565 ,

Unfortunately, `ulfius_add_endpoint_by_val` doesn't work with files in a folder, it requires a callback that will execute the code.

Also, you can't pass a regexp in the `url_format` parameter, but the `*` parameter is allowed as mentioned in the [documentation](https://github.com/babelouest/ulfius/blob/master/API.md#endpoint-structure-):
```
on an url_format that ends with '*', the rest of the url will not be tested
```

If you must use ulfius to provide static files, instead of using an apache or NGINX server instead, you can create a callback function that will behave like a static file reader though.
For that, you can use inspiration from the [example callback folder](https://github.com/babelouest/ulfius/tree/master/example_callbacks), like the [simple static file](https://github.com/babelouest/ulfius/tree/master/example_callbacks/static_file) or the [static file compressed output](https://github.com/babelouest/ulfius/tree/master/example_callbacks/static_compressed_inmemory_website) callbacks.

Be careful, they may be overkill as is if your needs are more simple than that.

In your case, a simple endpoint could start like this:

```C
ulfius_add_endpoint_by_val(&instance, "GET", "file", "*", 0, &callback_file, NULL);

int callback_file (const struct _u_request * request, struct _u_response * response, void * user_data) {
  char path[1025] = {0}, ;
  strcpy(path, "/path/to/directory/");
  strcat(path, request->http_url, 1024-strlen("/path/to/directory/"));

  FILE * f = fopen (path, "rb"); // be careful to check for directory traversal bugs
  //...
}
```
